sdk/lib/ksanitize/ubsan.c

```diff
@@ -0,0 +1,1775 @@
+}
+
+void
+__ubsan_handle_divrem_overflow(struct COverflowData *pData, unsigned long ulLHS, unsigned long ulRHS)
+{
+
+    ASSERT(pData);
+
+    HandleOverflow(false, pData, ulLHS, ulRHS, DIVREM_STRING);
+}
+
+void
+__ubsan_handle_divrem_overflow_abort(struct COverflowData *pData, unsigned long ulLHS, unsigned long ulRHS)
+{
+
+    ASSERT(pData);
+
+    HandleOverflow(true, pData, ulLHS, ulRHS, DIVREM_STRING);
+}
+
+void
+__ubsan_handle_dynamic_type_cache_miss(struct CDynamicTypeCacheMissData *pData, unsigned long ulPointer, unsigned long ulHash)
+{
+
+    ASSERT(pData);
+
+    HandleDynamicTypeCacheMiss(false, pData, ulPointer, ulHash);
+}
+
+void
+__ubsan_handle_dynamic_type_cache_miss_abort(struct CDynamicTypeCacheMissData *pData, unsigned long ulPointer, unsigned long ulHash)
+{
+
+    ASSERT(pData);
+
+    HandleDynamicTypeCacheMiss(false, pData, ulPointer, ulHash);
+}
+
+void
+__ubsan_handle_float_cast_overflow(struct CFloatCastOverflowData *pData, unsigned long ulFrom)
+{
+
+    ASSERT(pData);
+
+    HandleFloatCastOverflow(false, pData, ulFrom);
+}
+
+void
+__ubsan_handle_float_cast_overflow_abort(struct CFloatCastOverflowData *pData, unsigned long ulFrom)
+{
+
+    ASSERT(pData);
+
+    HandleFloatCastOverflow(true, pData, ulFrom);
+}
+
+void
+__ubsan_handle_function_type_mismatch(struct CFunctionTypeMismatchData *pData, unsigned long ulFunction)
+{
+
+    ASSERT(pData);
+
+    HandleFunctionTypeMismatch(false, pData, ulFunction);
+}
+
+void
+__ubsan_handle_function_type_mismatch_abort(struct CFunctionTypeMismatchData *pData, unsigned long ulFunction)
+{
+
+    ASSERT(pData);
+
+    HandleFunctionTypeMismatch(true, pData, ulFunction);
+}
+
+void
+__ubsan_handle_function_type_mismatch_v1(struct CFunctionTypeMismatchData *pData, unsigned long ulFunction, unsigned long ulCalleeRTTI, unsigned long ulFnRTTI)
+{
+
+    ASSERT(pData);
+#if 0
+    /*
+     * Unimplemented.
+     *
+     * This UBSan handler is special as the check has to be impelemented
+     * in an implementation. In order to handle it there is need to
+     * introspect into C++ ABI internals (RTTI) and use low-level
+     * C++ runtime interfaces.
+     */
+
+    HandleFunctionTypeMismatch(false, pData, ulFunction);
+#endif
+}
+
+void
+__ubsan_handle_function_type_mismatch_v1_abort(struct CFunctionTypeMismatchData *pData, unsigned long ulFunction, unsigned long ulCalleeRTTI, unsigned long ulFnRTTI)
+{
+
+    ASSERT(pData);
+#if 0
+    /*
+     * Unimplemented.
+     *
+     * This UBSan handler is special as the check has to be impelemented
+     * in an implementation. In order to handle it there is need to
+     * introspect into C++ ABI internals (RTTI) and use low-level
+     * C++ runtime interfaces.
+     */
+
+    HandleFunctionTypeMismatch(true, pData, ulFunction);
+#endif
+}
+
+void
+__ubsan_handle_invalid_builtin(struct CInvalidBuiltinData *pData)
+{
+
+    ASSERT(pData);
+
+    HandleInvalidBuiltin(true, pData);
```

Comment:
This indirection seems kind of unnecessary. Is this also taken from NetBSD or did you have to add it for some reason.

Reply:
No, I took it like this from NetBSD. The only things I changed are headers, `Report` function, and `%zu` specifier to `%Id`